In [107]:
import pandas as pd
import numpy as np
import seaborn as sns

In [165]:
x=pd.read_csv("Close_Date.csv")

In [109]:
x.columns

Index(['Zone', 'Site Address', 'Region', 'Opportunity ID', 'Opp Created On',
       'Enquiry Created By', 'Customer Name', 'Contact Number', 'E-Mail',
       'City', 'State', 'Pincode', 'Source of Enq', 'Sales Owner',
       'Sales Owner ID', 'CSE Name', 'Nature of Work', 'Influencer Involved',
       'Influencer Name', 'First Interaction Date', 'Site Visit Date',
       'Persuasion Date', 'Showroom Visit Date', 'Boq Received Date',
       'Product Category', 'Client expectation Price Bracket', 'Quote Segment',
       'WM Order Type', 'WM Quote ID', 'First Quote Date', 'Latest Quote Date',
       'Quote Review Date', 'Expected Closure Month', 'No. of Sales Quote',
       'Won/Lost/In Process', 'Last Progression Stage', 'Order in Hand',
       'Order in Hand Date', 'Status-Hot / Warm/ Cold', 'Remarks',
       'Main Competitor', 'Reason for Status', 'Expected Site Readiness Date',
       '1st Discount Given Date', '2nd Discount Given Date',
       'Order Basis to Fenesta', 'Quote Area', 

In [134]:
x.isna().sum()

Zone                     0
Site Address             0
Region                   0
Opportunity ID           0
Opp Created On           0
                        ..
Closed Customer value    0
Closed Basic value       0
MSC Material value       0
MSC Gross Value          0
Lost Price               0
Length: 63, dtype: int64

In [166]:
x.tail()

Zone                                       Site Address Region  \
10147  Not assigned     Village Hyatpur Rurki / Navashehar 144532 / IN      #   
10148  Not assigned  Vipanchika house ,SN Junction / Trippunithura ...      #   
10149  Not assigned               collecterate road / Rewa 486001 / IN      #   
10150  Not assigned                                 karnal 132001 / IN      #   
10151  Not assigned                                  satna 485001 / IN      #   

       Opportunity ID Opp Created On Enquiry Created By      Customer Name  \
10147          512479     14.02.2022     Connect Center      Amarjit Singh   
10148          533232     07.04.2022     Connect Center      Dr Sreejith S   
10149          552356     30.05.2022     Connect Center      Mohd Ishtiyaq   
10150          563350     24.06.2022              Sales    The Design Room   
10151          396128     19.03.2021                  #  Himansh Manghnani   

       Contact Number                          E-Mail           City  ...  \
10147  +91 8847569067  Sarvjitkaur_kaur2132@yahoo.com     Navashehar  ...   
10148  +91 9400626123              drsreeje@gmail.com  Trippunithura  ...   
10149  +91 9425841888         ishtiyaq.rewa@gmail.com           Rewa  ...   
10150  +91 9518069000            veekayguru@gmail.com         karnal  ...   
10151  +91 8878886488   himanshumanghnani01@gmail.com          satna  ...   

          Gross Value       Net Value Basic Value (at DP) Standard Discount  \
10147               0               0                   0                 0   
10148               0               0                   0                 0   
10149               0               0                   0                 0   
10150  169,962.29 INR  137,901.10 INR      124,110.99 INR     18,804.70 INR   
10151               0               0                   0                 0   

      1st Discount Given Closed Customer value Closed Basic value  \
10147           0.00 INR                     0                  0   
10148           0.00 INR                     0                  0   
10149           0.00 INR                     0                  0   
10150           0.00 INR              0.00 INR           0.00 INR   
10151           0.00 INR                     0                  0   

      MSC Material value MSC Gross Value Lost Price  
10147                  0               0          0  
10148                  0               0          0  
10149                  0               0          0  
10150           0.00 INR               0          0  
10151                  0               0          0  

[5 rows x 63 columns]

In [167]:
x.replace("#",np.NaN,inplace=True)

In [168]:
x.tail()

Zone                                       Site Address Region  \
10147  Not assigned     Village Hyatpur Rurki / Navashehar 144532 / IN    NaN   
10148  Not assigned  Vipanchika house ,SN Junction / Trippunithura ...    NaN   
10149  Not assigned               collecterate road / Rewa 486001 / IN    NaN   
10150  Not assigned                                 karnal 132001 / IN    NaN   
10151  Not assigned                                  satna 485001 / IN    NaN   

       Opportunity ID Opp Created On Enquiry Created By      Customer Name  \
10147          512479     14.02.2022     Connect Center      Amarjit Singh   
10148          533232     07.04.2022     Connect Center      Dr Sreejith S   
10149          552356     30.05.2022     Connect Center      Mohd Ishtiyaq   
10150          563350     24.06.2022              Sales    The Design Room   
10151          396128     19.03.2021                NaN  Himansh Manghnani   

       Contact Number                          E-Mail           City  ...  \
10147  +91 8847569067  Sarvjitkaur_kaur2132@yahoo.com     Navashehar  ...   
10148  +91 9400626123              drsreeje@gmail.com  Trippunithura  ...   
10149  +91 9425841888         ishtiyaq.rewa@gmail.com           Rewa  ...   
10150  +91 9518069000            veekayguru@gmail.com         karnal  ...   
10151  +91 8878886488   himanshumanghnani01@gmail.com          satna  ...   

          Gross Value       Net Value Basic Value (at DP) Standard Discount  \
10147               0               0                   0                 0   
10148               0               0                   0                 0   
10149               0               0                   0                 0   
10150  169,962.29 INR  137,901.10 INR      124,110.99 INR     18,804.70 INR   
10151               0               0                   0                 0   

      1st Discount Given Closed Customer value Closed Basic value  \
10147           0.00 INR                     0                  0   
10148           0.00 INR                     0                  0   
10149           0.00 INR                     0                  0   
10150           0.00 INR              0.00 INR           0.00 INR   
10151           0.00 INR                     0                  0   

      MSC Material value MSC Gross Value Lost Price  
10147                  0               0          0  
10148                  0               0          0  
10149                  0               0          0  
10150           0.00 INR               0          0  
10151                  0               0          0  

[5 rows x 63 columns]

In [139]:
x.isna().sum()

Zone                      0
Site Address              0
Region                   95
Opportunity ID            0
Opp Created On            0
                         ..
Closed Customer value     0
Closed Basic value        0
MSC Material value        0
MSC Gross Value           0
Lost Price                0
Length: 63, dtype: int64

In [114]:
#checking all cities in region column 
x['Region'].unique()

array(['UPC BLR 2', 'Bangalore City', 'UPC BLR 1', 'Karnataka-Upcountry',
       'Bangalore', 'ROTN', 'Chennai', 'Chennai City', 'Kerala',
       'Central Bengal', 'Kolkata City', 'North East', 'East-Zone',
       'Kolkata', nan, 'North Bengal', 'Jharkhand', 'East Central',
       'Chattisgarh', 'Bihar', 'Odisha', 'Orissa', 'Patna', 'Raipur',
       'Gurgaon', 'HO', 'Hyderabad City', 'Andhra Pradesh', 'Telanagana',
       'Hyderabad', 'hyderabad', 'BILASPUR', 'NCR 2', 'NCR 1', 'NCR',
       'NCR 3', 'Faridabad', 'NCR2', 'Delhi', 'PB', 'UT', 'EUP', 'CUP',
       'Chandigarh', 'Rohtak', 'Uttar Pradesh - East', 'North-Zone',
       'Punjab', 'Pune - City', 'Pune', 'Nagpur', 'MMR 2', 'MMR 1', 'MMR',
       'Goa', 'Mumbai', 'Kolhapur', 'Rajasthan', 'Gujarat',
       'Madhya Pradesh', 'Jaipur', 'BHOPAL'], dtype=object)

In [140]:
#checking how much time cities repited  in region column 
x.groupby('Region')['Region'].agg("count")

Region
Andhra Pradesh          380
BHOPAL                   10
BILASPUR                  1
Bangalore                52
Bangalore City          534
                       ... 
UPC BLR 1               289
UPC BLR 2               146
UT                      202
Uttar Pradesh - East      5
hyderabad                 2
Name: Region, Length: 61, dtype: int64

In [141]:
# ckecking for blank columns      
x.isna().sum()[1:10]                      

Site Address            0
Region                 95
Opportunity ID          0
Opp Created On          0
Enquiry Created By    231
Customer Name           0
Contact Number          4
E-Mail                  4
City                    0
dtype: int64

In [142]:
print("no. of rows :-   ",x.shape[0])
print("no. of columns:- ",x.shape[-1])


no. of rows :-    10152
no. of columns:-  63


#  null values in column 
Created By    231         fill
CSE Name                  1994                          #fill
Nature of Work             145                          #ratio check fill
Influencer Name           8538   (84% unavailable data) #fill              
First Interaction Date    1334                 -------#filling date
Site Visit Date                     5366              ""                  Type-date
Persuasion Date                     8289          #drop (81% data unavailable)     ""
Showroom Visit Date                 7323             ""
Boq Received Date                   7725               ""
Order in Hand              10146             drop
Order in Hand Date         10146              drop
Latest Quote Date           2027                       fill   Type-date         
Quote Review Date           9478              #drop
Expected Site Readiness Date      968                    ""
1st Discount Given Date          9672         #drop
2nd Discount Given Date         10077         #drop  
Order Basis to Fenesta           3497                 fill           
Quote Area                       3497                 mean fill
Next Step                        8443      /drop/fill? (83% unavailable data)
Next Step Due Date               9837  drop
Region                 95                             fill
Enquiry Created By    231                  fill
Sales Owner ID        74    mean 

In [189]:
#checking null values in specific column
x["Nature of Work"].isna().sum()


0

In [224]:
x["Nature of Work"][20:44]     #ratio check

20     New Built
21     New Built
22     New Built
23     New Built
24     New Built
25    Renovation
26     New Built
27     New Built
28     New Built
29    Renovation
30     New Built
31     New Built
32     New Built
33     New Built
34    Renovation
35    Renovation
36     New Built
37    Renovation
38     New Built
39     New Built
40    Renovation
41    Renovation
42     New Built
43     New Built
Name: Nature of Work, dtype: object

In [169]:
x['Region']=x['Region'].ffill()    # region fill

In [178]:
x["CSE Name"].isna().sum()

0

In [177]:
x['CSE Name']=x['CSE Name'].ffill() # CSE name fill

In [199]:
# changing data type of (Natue of work) column
x["Nature of Work"].info()

<class 'pandas.core.series.Series'>
RangeIndex: 10152 entries, 0 to 10151
Series name: Nature of Work
Non-Null Count  Dtype 
--------------  ----- 
10007 non-null  object
dtypes: object(1)
memory usage: 79.4+ KB


In [213]:
#changing data type of (nature of work)
x["Nature of Work"]=x["Nature of Work"].astype("object")

In [214]:
a=0                                   #ratio check
b=0
for i in x["Nature of Work"]:
    if 'New Built' in i:
        a=a+1
    else:
        b=b+1
print("New Built:- ",a)
print("Renovation:-",b)

New Built:-  6779
Renovation:- 3373


In [222]:
x["Nature of Work"]=x["Nature of Work"].replace("nan","New Built",regex=True)

In [279]:
x["Influencer Name"].isna().sum()

8538

In [4]:
x["1st Discount Given Date"].isna().sum()

NameError: name 'x' is not defined

In [3]:
x.drop(columns=['Enquiry Created By', 'Persuasion Date','Showroom Visit Date'])

NameError: name 'x' is not defined

In [280]:
x.pop("Influencer Name")

0        Arc Parithosh Parithosh Architects
1                                       NaN
2                                       NaN
3                                       NaN
4                                       NaN
                        ...                
10147                                   NaN
10148                                   NaN
10149                                   NaN
10150                                   NaN
10151                                   NaN
Name: Influencer Name, Length: 10152, dtype: object

In [281]:
x.shape

(10152, 52)

In [288]:
x['Order Basis to Fenesta']=x['Order Basis to Fenesta'].ffill()

In [295]:
x['Quote Area'].info()

<class 'pandas.core.series.Series'>
RangeIndex: 10152 entries, 0 to 10151
Series name: Quote Area
Non-Null Count  Dtype 
--------------  ----- 
6655 non-null   object
dtypes: object(1)
memory usage: 79.4+ KB


In [296]:
x['Quote Area'].isna().sum()

3497

In [298]:
x["Quote Area"]

0         110.24
1        1132.22
2         569.12
3         111.85
4          56.85
          ...   
10147        NaN
10148        NaN
10149        NaN
10150     165.53
10151        NaN
Name: Quote Area, Length: 10152, dtype: object

In [299]:
x["Quote Area"]=x["Quote Area"].astype(float)

In [300]:
x["Quote Area"].info()

<class 'pandas.core.series.Series'>
RangeIndex: 10152 entries, 0 to 10151
Series name: Quote Area
Non-Null Count  Dtype  
--------------  -----  
6655 non-null   float64
dtypes: float64(1)
memory usage: 79.4 KB


In [301]:
x["Quote Area"].mean()

732.4408069120963

In [302]:
x["Quote Area"].fillna(x["Quote Area"].mean(),inplace=True)

In [344]:
x.isna().sum()[50:511] 

Lost Price    0
dtype: int64

In [305]:
x['Contact Number']=x['Contact Number'].ffill()

In [306]:
x['E-Mail']=x['E-Mail'].ffill()

In [314]:
x["Sales Owner ID"].info()

<class 'pandas.core.series.Series'>
RangeIndex: 10152 entries, 0 to 10151
Series name: Sales Owner ID
Non-Null Count  Dtype  
--------------  -----  
10078 non-null  float64
dtypes: float64(1)
memory usage: 79.4 KB


In [313]:
x["Sales Owner ID"]=x["Sales Owner ID"].astype(float)

In [315]:
x["Sales Owner ID"].fillna(x["Sales Owner ID"].mean(),inplace=True)

In [335]:
x['WM Quote ID']=x['WM Quote ID'].bfill()

In [338]:
x['Remarks']=x['Remarks'].bfill()

In [327]:
x.pop("Order in Hand Date")

0        NaN
1        NaN
2        NaN
3        NaN
4        NaN
        ... 
10147    NaN
10148    NaN
10149    NaN
10150    NaN
10151    NaN
Name: Order in Hand Date, Length: 10152, dtype: object

In [339]:
x["Remarks"]

0        They contacted 10 days ago. Now they don't hav...
1        They contacted 10 days ago. Now they don't hav...
2        They contacted 10 days ago. Now they don't hav...
3        They contacted 10 days ago. Now they don't hav...
4        They contacted 10 days ago. Now they don't hav...
                               ...                        
10147         CALL DONE. WILL VISIT CLIENT SITE NEXT WEEK.
10148                                   HOSPITAL EMERGENCY
10149    Had word with client need window for renovated...
10150    Had word with client need window for renovated...
10151                        client use the other material
Name: Remarks, Length: 10152, dtype: object

In [345]:
x.isna().sum()

Zone                                0
Site Address                        0
Region                              0
Opportunity ID                      0
Opp Created On                      0
Customer Name                       0
Contact Number                      0
E-Mail                              0
City                                0
State                               0
Pincode                             0
Source of Enq                       0
Sales Owner                         0
Sales Owner ID                      0
CSE Name                            0
Nature of Work                      0
Influencer Involved                 0
First Interaction Date              0
Site Visit Date                     0
Product Category                    0
Client expectation Price Bracket    0
Quote Segment                       0
WM Order Type                       0
WM Quote ID                         0
First Quote Date                    0
Latest Quote Date                   0
Expected Clo

In [349]:
x.to_excel('1a close data.xlsx')